### ISSUE TITLE
Unexpected result for pygeos.get_coordinates

### ISSUE THREAD

**Issue by sziem** [shapely/shapely#1262]
This is an issue with pygeos, which I am raising here as requested in the note over there.

## Expected behavior and actual behavior.

The result of get_coordinates is different from my expectation.

If I create linestrings from two lists of coordinates, I get two linestrings as expected. However, when I run get_coordinates, both linestrings are concatenated into a single array.

## Steps to reproduce the problem.
```python
import pygeos
linestrings = pygeos.linestrings([[[0, 0], [0, 1]], [[3, 3], [3, 4]]]).tolist()
# returns a list of 2 linestrings as expected
# [<pygeos.Geometry LINESTRING (0 0, 0 1)>, <pygeos.Geometry LINESTRING (3 3, 3 4)>]
pygeos.get_coordinates(linestrings).tolist()
# returns a single list.
# [[0.0, 0.0], [0.0, 1.0], [3.0, 3.0], [3.0, 4.0]]
# My expectation is to get 2 lists (that is, the input list again).
# [[[0, 0], [0, 1]], [[3, 3], [3, 4]]]
```

## Operating system

Manjaro Linux
Release:	21.2rc

## pygeos version
```
pygeos.__version__ 
# '0.12.0'
```

**Comment by jorisvandenbossche** [comment]
As you mention, it is the documented behaviour of `get_coordinates` to return a "flat" array of all coordinates. There is also a `return_index=True` option so you can keep track of which coordinates are from which geometry.

The main problem is that your expected behaviour of two lists doesn't work with numpy arrays (which requires constant dimensions). 
But using the `return_index=True` option, you can achieve what you want:

```python
In [12]: linestrings = shapely.linestrings([[[0, 0], [0, 1]], [[3, 3], [3, 4]]]).tolist()
    ...: coords, indices = shapely.get_coordinates(linestrings, return_index=True)

In [13]: coords
Out[13]: 
array([[0., 0.],
       [0., 1.],
       [3., 3.],
       [3., 4.]])

In [14]: indices
Out[14]: array([0, 0, 1, 1])

In [15]: [coords[indices == i].tolist() for i in range(len(linestrings))]
Out[15]: [[[0.0, 0.0], [0.0, 1.0]], [[3.0, 3.0], [3.0, 4.0]]]
``` 

Or probably easier, you can start with such a list comprehension when calling `get_coordinates`:

```python
In [16]: [shapely.get_coordinates(line).tolist() for line in linestrings]
Out[16]: [[[0.0, 0.0], [0.0, 1.0]], [[3.0, 3.0], [3.0, 4.0]]]
```

**Comment by sziem** [comment]
Thanks for the suggestion with `return_index`.

> The main problem is that your expected behaviour of two lists doesn't work with numpy arrays (which requires constant dimensions).

I get that point.  However, a lot of the pygeos functions operate on object arrays of lists, for example the line

**Comment by sziem** [comment]
I have just started using pygeos and this came to my mind.  If you think that the current behaviour fits well into the interface, I can close this.

**Comment by jorisvandenbossche** [comment]
> However, a lot of the pygeos functions return object arrays of lists

I don't think this is the case (do you have such an example where we return an object array of lists?). We _do_ return (and accept as input) object-dtype arrays with geometry objects, but not with coordinates.

**Comment by sziem** [comment]
You're right, the example I mentioned returns an object array with geometry objects.  Apologies for that.  What do you think about the second suggestion, to return some kind of list of arrays?

**Comment by jorisvandenbossche** [comment]
You can relatively easily do this with `[shapely.get_coordinates(line).tolist() for line in linestrings]` as mentioned above, or even simpler with `[list(line.coords) for line in linestrings]`. 
Given that, I am not sure it is needed to add a function for this (we can certainly document this better though, and add such example to the docs)

**Comment by jbbqqf** [comment]
Hi @sziem — drive-by triage of stale issues: I believe this one can be closed.

Re-reading the thread, you offered to close it yourself in [your 2021-12-16 comment](https://github.com/shapely/shapely/issues/1262#issuecomment-996187960) ("If you think that the current behaviour fits well into the interface, I can close this."), and @jorisvandenbossche [declined the feature add](https://github.com/shapely/shapely/issues/1262#issuecomment-997280466) the next day, suggesting the two one-line workarounds:

```python
[shapely.get_coordinates(line).tolist() for line in linestrings]
# or
[list(line.coords) for line in linestrings]
```

The behavior is unchanged in current shapely 2.1: `shapely.get_coordinates` still returns a flat array with the documented `return_index=True` option for splitting back out per-geometry. Documentation improvement (the only side-action the member mentioned) is orthogonal and would be a separate PR/issue.

Probably safe to close. Happy to be corrected if I'm missing something.

---
_Drive-by triage from an outside contributor (data engineer GCP/Python, [jbbqqf](https://github.com/jbbqqf)). Comment drafted with Claude assistance; evidence and conclusion are mine._
